[AIRFLOW-2985] Operators for S3 object copying/deleting

XD-DENG opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2985
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Currently we don't have an operator in Airflow to help copy objects within S3, while this is a quite common use case when we deal with the data in S3.

Under the hood, this operator is using `boto3.client.copy_object()`.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Test case has been added

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

ashb commented on airflow/contrib/operators/s3_copy_object_operator.py:
```diff
@@ -0,0 +1,84 @@
+            self,
+            dest_s3_bucket,
+            dest_s3_key,
+            source_s3_bucket,
```
It would be nice to support a single `s3://bucket/key` style parameter here like we do in the other S3 ops/sensors https://github.com/apache/incubator-airflow/blob/ac9033db0981ae1f770a8bdb5597055751ab15bd/airflow/sensors/s3_key_sensor.py#L34-L39

ashb commented on airflow/contrib/operators/s3_copy_object_operator.py:
```diff
@@ -0,0 +1,84 @@
+from airflow.utils.decorators import apply_defaults
+
+
+class S3CopyObjectOperator(BaseOperator):
```
Please add a link to this class in docs/code.rst

ashb commented on airflow/contrib/operators/s3_copy_object_operator.py:
```diff
@@ -0,0 +1,129 @@
+    def __init__(
+            self,
+            dest_bucket_key,
+            source_bucket_key,
```
Let's swap these two params around so that it mirrors `cp src dest`

XD-DENG replied:
Good point. Will change accordingly.

ashb commented on airflow/contrib/operators/s3_copy_object_operator.py:
```diff
@@ -0,0 +1,129 @@
+                      'VersionId': self.source_version_id}
+        response = s3_hook.get_conn().copy_object(Bucket=self.dest_bucket_name,
+                                                  Key=self.dest_bucket_key,
+                                                  CopySource=CopySource)
```
I wonder if this should actually live in the S3 hook? Thoughts?

XD-DENG replied:
I agree we can have `copy_object()` and `delete_object()` as methods of S3Hook(), while it would still be necessary to have two operators as wrppers of these two methods. Do you agree?

(i.e., we put the logics  into S3Hook() as methods, and then invoke them in the operators)

ashb replied:
Yes, logic in S3 hook, two operators.

ashb commented on airflow/contrib/operators/s3_copy_object_operator.py:
```diff
@@ -0,0 +1,129 @@
+            if parsed_url.scheme != '' or parsed_url.netloc != '':
+                raise AirflowException('If dest_bucket_name is provided, ' +
+                                       'dest_bucket_key should be relative path ' +
+                                       'from root level, rather than a full s3:// url')
```
A lot of this is duplicated between src and dest, and already duplicates what lives in the `S3Hook.parse_s3_url`.

We should at least use that method (it's static, so we don't have to create a hook instance in the constructor).

XD-DENG replied:
Sorry I didn't realise there is a `parse_s3_url()` method in `S3Hook()`.

Will change accordingly.

ashb commented on airflow/contrib/operators/s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,88 @@
+                                                     Delete=delete_dict)
+
+        deleted_keys = [x['Key'] for x in response.get("Deleted", [])]
+        errors_keys = [x['Key'] for x in response.get("Errors", [])]
```
I think all this logic should live in the S3Hook, not in this operator.

XD-DENG replied:
Will change accordingly.

ashb commented on airflow/contrib/operators/s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,88 @@
+        errors_keys = [x['Key'] for x in response.get("Errors", [])]
+        self.log.info("Deleted: {}".format(deleted_keys))
+        self.log.info("Errors: {}".format(errors_keys))
+        self.log.info("Full Response:\n{}".format(pformat(response)))
```
Do we want to log the full response (espeically at info) - that seems verbose.

XD-DENG replied:
Normally the response here is not that long. But yes, putting them in info may not be a good choice.

Do you think whether there is any other logging "location" that may be more proper for this? If not, I'm delete this full response logging.

ashb replied:
Probably just remove this.

ashb commented on airflow/contrib/operators/s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,88 @@
+        deleted_keys = [x['Key'] for x in response.get("Deleted", [])]
+        errors_keys = [x['Key'] for x in response.get("Errors", [])]
+        self.log.info("Deleted: {}".format(deleted_keys))
+        self.log.info("Errors: {}".format(errors_keys))
```
If there are errors then then task should fail.

XD-DENG replied:
`Errors` here appear only because users specified some keys to delete while these keys don't exist in the bucket. It doesn't have to mean an actual error.

If we fail on "error" here, implicitly we're adding a requirement: you can only delete objects that you're 100% sure that they exist in the bucket.

Or possibly we can add another argument, say `fail_if_not_found`. Only when it's set to `True`, the task would fail when the key to delete doesn't exist. (default value should be?)

Any thoughts on this?

ashb replied:
Ah right, if that's the only time it can fail then this is the right behaviour - failing to delete a file that doesn't exist is not an error, but what about other cases, such as failing to delete because the object is owned by another user (something I run across frequently with cross-account bucket set ups))

XD-DENG replied:
I believe the case you described would return an exception instead, while the `Errors` I mention here is an entry in the response from a successful request.

On the other hand, I'll double-confirm whether `failing to delete a file that doesn't exist` is the only case causing `Errors` entry in the response from a successful request.

ashb commented on airflow/contrib/operators/s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,88 @@
+        if isinstance(keys, list):
+            self.keys = keys
+        else:
+            self.keys = [keys]
```
It might be nice to support the `s3://` style in the single key case?

ashb replied:
(Maybe. I'm not sure if supporting in this case but not the others is more or less confusing.)

XD-DENG replied:
Personally I still suggest not support the `s3://` style for `S3DeleteObjectsOperator`, as it makes the argument combination more complicated, which may lead to another sort of confusion.

XD-DENG commented on airflow/contrib/operators/s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,92 @@
+        deleted_keys = [x['Key'] for x in response.get("Deleted", [])]
+        self.log.info("Deleted: {}".format(deleted_keys))
+
+        if not self.silent_on_errors and "Errors" in response:
```
Regarding the `Errors` in the boto3 response, I use argument `silent_on_errors` to enable users decide whether they consider `deleting non-existent object` as exceptions and if the operator should fail.

XD-DENG commented on airflow/hooks/S3_hook.py:
```diff
@@ -384,3 +384,89 @@ def load_bytes(self,
         client = self.get_conn()
         client.upload_fileobj(filelike_buffer, bucket_name, key, ExtraArgs=extra_args)
+
+    def _copy_object(self,
```
I name the methods in `S3Hook()` as `_copy_object`, in order to avoid confusion between `boto3.client.copy_object` and `S3Hook.copy_object`.

The same applies to  `_delete_objects`

ashb replied:
I wouldn't. A method prefixed with `_` in python is usually an indication that it is "private" and shouldn't be called from outside the method.

Given the S3Hook doesn't directly expose any methods from the boto client/session object I think the chance of confusion is slight. (and a doc string highlighting any differences will help dispel confusion)

XD-DENG replied:
Sure, I will change this part.

Actually the chance of misusing is low given the argument required are totally different.

XD-DENG commented on tests/contrib/operators/test_s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,111 @@
+                                                         Prefix=key_pattern))
+
+    @mock_s3
+    def test_s3_delete_non_existent_object(self):
```
A test case is added to test the argument `silent_on_errors`.

XD-DENG commented on airflow/contrib/operators/s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,92 @@
+from airflow.utils.decorators import apply_defaults
+
+
+class S3DeleteObjectsOperator(BaseOperator):
```
Personally I still suggest not support S3 style url in `S3DeleteObjectsOperator`. This is to make
- argument combination clear
- support deleting single object and deleting multiple object within one operator.

XD-DENG commented on airflow/hooks/S3_hook.py:
```diff
@@ -384,3 +384,89 @@ def load_bytes(self,
+        response = self.get_conn().copy_object(Bucket=dest_bucket_name,
+                                               Key=dest_bucket_key,
+                                               CopySource=CopySource)
+        return response
```
For both `copy_object()` and `delete_objects()`, I'm returning the response from boto3/S3 API directly. This can give users the flexibility if they want to dive deeper into the details in the response.

XD-DENG commented on airflow/contrib/operators/s3_copy_object_operator.py:
```diff
@@ -0,0 +1,96 @@
+            source_bucket_name=None,
+            dest_bucket_name=None,
+            source_version_id=None,
+            aws_conn_id='aws_default',
```
Earlier this argument was named `s3_conn_id`. I change it to `aws_conn_id` to make it consistent with most of the S3 operators/sensors.

The same chagne is made in `S3DeleteObjectsOperator` as well.

XD-DENG committed:
[AIRFLOW-2985] Operators for S3 object copying/deleting

1. Copying:
Under the hood, it's `boto3.client.copy_object()`.
It can only handle the situation in which the
S3 connection used can access both source and
destination bucket/key.

2. Deleting:
2.1 Under the hood, it's `boto3.client.delete_objects()`.
It supports either deleting one single object or
multiple objects.
2.2 If users try to delete a non-existent object, the
request will still succeed, but there will be an
entry 'Errors' in the response. There may also be
other reasons which may cause similar 'Errors' (
request itself would succeed without explicit
exception). So an argument `silent_on_errors` is added
to let users decide if this sort of 'Errors' should
fail the operator.

The corresponding methods are added into S3Hook, and
these two operators are 'wrappers' of these methods.

XD-DENG commented on tests/contrib/operators/test_s3_delete_objects_operator.py:
```diff
@@ -0,0 +1,85 @@
+
+        # There should be no object found in the bucket created earlier
+        self.assertFalse('Contents' in conn.list_objects(Bucket=bucket,
+                                                         Prefix=key_pattern))
```
I deleted test case `test_s3_delete_non_existent_object`. It can pass CI, but I realize it can only be reproduced by `moto.mock_s3`.

That is, when we use `moto.mock_s3` to mirror the S3 environment, deleting an non-existent object would result in an `Errors` entry in the response. But in AWS S3 and IBM COS (S3-compatible), deleting non-existent object is allowed, and the non-existent key would still be in the `Deleted` entry.

**Regarding `S3DeleteObjectsOperator()`, if there is `Errors` entry in the response, the operator will raise exception and fail, as you advised earlier.**

**Reference**:
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.delete_objects
Response Syntax
```
{
    'Deleted': [
        {
            'Key': 'string',
            'VersionId': 'string',
            'DeleteMarker': True|False,
            'DeleteMarkerVersionId': 'string'
        },
    ],
    'RequestCharged': 'requester',
    'Errors': [
        {
            'Key': 'string',
            'VersionId': 'string',
            'Code': 'string',
            'Message': 'string'
        },
    ]
}
```

ashb replied:
Confirmed, sounds good.
